File: python/musica/tuvx/radiator.py
# Copyright (C) 2023-2026 University Corporation for Atmospheric Research
# SPDX-License-Identifier: Apache-2.0
"""
TUV-x Radiator class.

This module provides a class for defining radiators in TUV-x. Radiators represent
optically active species that should be considered in radiative transfer calculations.

Note: TUV-x is only available on macOS and Linux platforms.
"""

from typing import Optional
import numpy as np
from .. import backend
from .._base import CppWrapper, _unwrap
from .grid import Grid

_backend = backend.get_backend()
_Radiator = _backend._tuvx._Radiator if backend.tuvx_available() else None


class Radiator(CppWrapper):
    """An optically active species for TUV-x radiative transfer calculations.

    Attributes:
        name: Name of the radiator.
        number_of_height_sections: Number of height grid sections.
        number_of_wavelength_sections: Number of wavelength grid sections.
        optical_depths: 2D array of optical depths (zero-copy).
        single_scattering_albedos: 2D array of single scattering albedos (zero-copy).
        asymmetry_factors: 2D array of asymmetry parameters (zero-copy).
    """

    _unavailable_message = "TUV-x was not included in your build of MUSICA."

    @classmethod
    def _check_available(cls):
        return backend.tuvx_available()

    def __init__(self, *, name: str, height_grid: Grid, wavelength_grid: Grid,
                 optical_depths: Optional[np.ndarray] = None,
                 single_scattering_albedos: Optional[np.ndarray] = None,
                 asymmetry_factors: Optional[np.ndarray] = None,
                 **kwargs):
        """Initialize a Radiator instance.

        Args:
            name: Name of the radiator.
            height_grid: Grid on which the radiator is defined (height).
            wavelength_grid: Grid on which the radiator is defined (wavelength).
            optical_depths: Optional 2D array of optical depths
                            (shape: num_wavelengths x num_heights).
            single_scattering_albedos: Optional 2D array of single scattering albedos
                                       (shape: num_wavelengths x num_heights).
            asymmetry_factors: Optional 2D array of asymmetry parameters
                               (shape: num_wavelengths x num_heights).
            **kwargs: Additional arguments passed to the C++ constructor.
        """
        if not self._check_available():
            raise RuntimeError(self._unavailable_message)
        self._cpp = _Radiator(name=name, height_grid=_unwrap(height_grid),
                              wavelength_grid=_unwrap(wavelength_grid), **kwargs)
        if optical_depths is not None:
            self.optical_depths = optical_depths
        if single_scattering_albedos is not None:
            self.single_scattering_albedos = single_scattering_albedos
        if asymmetry_factors is not None:
            self.asymmetry_factors = asymmetry_factors

    @property
    def name(self) -> str:
        """Name of the radiator."""
        return self._cpp.name

    @name.setter
    def name(self, value: str):
        self._cpp.name = value

    @property
    def number_of_height_sections(self) -> int:
        """Number of height grid sections."""
        return self._cpp.number_of_height_sections

    @property
    def number_of_wavelength_sections(self) -> int:
        """Number of wavelength grid sections."""
        return self._cpp.number_of_wavelength_sections

    @property
    def optical_depths(self) -> np.ndarray:
        """2D array of optical depths (zero-copy view into C++ memory)."""
        return self._cpp.optical_depths

    @optical_depths.setter
    def optical_depths(self, value: np.ndarray):
        self._cpp.optical_depths = value

    @property
    def single_scattering_albedos(self) -> np.ndarray:
        """2D array of single scattering albedos (zero-copy view into C++ memory)."""
        return self._cpp.single_scattering_albedos

    @single_scattering_albedos.setter
    def single_scattering_albedos(self, value: np.ndarray):
        self._cpp.single_scattering_albedos = value

    @property
    def asymmetry_factors(self) -> np.ndarray:
        """2D array of asymmetry parameters (zero-copy view into C++ memory)."""
        return self._cpp.asymmetry_factors

    @asymmetry_factors.setter
    def asymmetry_factors(self, value: np.ndarray):
        self._cpp.asymmetry_factors = value

    def __str__(self):
        """User-friendly string representation."""
        return (f"Radiator(name={self.name}, "
                f"num_height_sections={self.number_of_height_sections}, "
                f"num_wavelength_sections={self.number_of_wavelength_sections})")

    def __repr__(self):
        """Detailed string representation for debugging."""
        return (f"Radiator(name={self.name}, "
                f"num_height_sections={self.number_of_height_sections}, "
                f"num_wavelength_sections={self.number_of_wavelength_sections}, "
                f"optical_depths={self.optical_depths}, "
                f"single_scattering_albedos={self.single_scattering_albedos}, "
                f"asymmetry_factors={self.asymmetry_factors})")

    def __eq__(self, other):
        """Check equality between two Radiator instances."""
        if not isinstance(other, Radiator):
            return NotImplemented
        return (self.name == other.name and
                np.array_equal(self.optical_depths, other.optical_depths) and
                np.array_equal(self.single_scattering_albedos, other.single_scattering_albedos) and
                np.array_equal(self.asymmetry_factors, other.asymmetry_factors) and
                self.number_of_height_sections == other.number_of_height_sections and
                self.number_of_wavelength_sections == other.number_of_wavelength_sections)

    def __bool__(self):
        """Return True if the radiator has name and height and wavelength sections."""
        return (bool(self.name) and
                self.number_of_height_sections > 0 and
                self.number_of_wavelength_sections > 0)

File: julia/src/tuvx/radiator.jl
# Copyright (C) 2023-2026 University Corporation for Atmospheric Research
# SPDX-License-Identifier: Apache-2.0

const RadiatorPtr = CxxWrap.CxxWrapCore.CxxPtr{CppRadiator}

"""
    RadiatorView <: AbstractMatrix{Float64}

A zero-copy view into a 2D value array of a [`Radiator`](@ref) (optical
depths, single scattering albedos, or asymmetry factors).

The array is stored height-fastest in memory, which is exactly Julia's
column-major layout for a `(num_height_sections, num_wavelength_sections)`
matrix, so no transpose is needed. Element `view[h, w]` is the value at
height section `h` and wavelength section `w`.

The view shares memory with the TUV-x radiator, so a write to the view
changes the radiator itself. The view also holds a reference to its radiator.
The reference stops the garbage collector from freeing the radiator while the
view is alive.

Get a new view after you add the radiator to a [`RadiatorMap`](@ref). The map
takes over the memory of the radiator, and an older view then points to
memory that is free.
"""
struct RadiatorView <: AbstractMatrix{Float64}
    _data::Matrix{Float64}
    _radiator::Any  # prevents GC of the parent Radiator while this view is alive
end

Base.size(view::RadiatorView) = size(view._data)
Base.IndexStyle(::Type{RadiatorView}) = IndexLinear()
Base.getindex(view::RadiatorView, i::Int) = view._data[i]
Base.getindex(view::RadiatorView, i::Int, j::Int) = view._data[i, j]
Base.setindex!(view::RadiatorView, value, i::Int) = (view._data[i] = value)
Base.setindex!(view::RadiatorView, value, i::Int, j::Int) = (view._data[i, j] = value)

"""
    Radiator

An optically active species for TUV-x radiative transfer calculations, such
as an aerosol layer.

# Constructor

    Radiator(; name, height_grid, wavelength_grid, optical_depths=nothing,
               single_scattering_albedos=nothing, asymmetry_factors=nothing)

- `name::AbstractString`: The name of the radiator
- `height_grid::Grid`: The height grid on which the radiator is defined
- `wavelength_grid::Grid`: The wavelength grid on which the radiator is defined
- `optical_depths::AbstractMatrix{<:Real}`: Optical depths, shape `(num_height_sections(height_grid), num_sections(wavelength_grid))`
- `single_scattering_albedos::AbstractMatrix{<:Real}`: Single scattering albedos, same shape as `optical_depths`
- `asymmetry_factors::AbstractMatrix{<:Real}`: Asymmetry factors, same shape as `optical_depths`

The number of streams is currently fixed at 1 in TUV-x, so asymmetry factors
are exposed as a 2D array.

# Example

```julia
height_grid = Grid(name = "height", units = "km", edges = [0.0, 2.0, 4.0])
wavelength_grid = Grid(name = "wavelength", units = "nm", edges = [200.0, 300.0])
radiator = Radiator(name = "aerosol", height_grid = height_grid, wavelength_grid = wavelength_grid)
```
"""
mutable struct Radiator
    _ptr::RadiatorPtr
    _owner::Any  # the RadiatorMap that owns the TUV-x radiator, or nothing

    function Radiator(ptr::RadiatorPtr, owner)
        obj = new(ptr, owner)
        finalizer(obj) do r
            cpp_delete_radiator(r._ptr)
        end
        return obj
    end
end

function Radiator(;
    name::AbstractString,
    height_grid::Grid,
    wavelength_grid::Grid,
    optical_depths::Union{AbstractMatrix{<:Real},Nothing} = nothing,
    single_scattering_albedos::Union{AbstractMatrix{<:Real},Nothing} = nothing,
    asymmetry_factors::Union{AbstractMatrix{<:Real},Nothing} = nothing,
)
    radiator = Radiator(
        cpp_create_radiator(String(name), height_grid._ptr, wavelength_grid._ptr),
        nothing,
    )

    optical_depths !== nothing && set_optical_depths!(radiator, optical_depths)
    single_scattering_albedos !== nothing &&
        set_single_scattering_albedos!(radiator, single_scattering_albedos)
    asymmetry_factors !== nothing && set_asymmetry_factors!(radiator, asymmetry_factors)

    return radiator
end

"""
    get_name(radiator::Radiator) -> String

Get the name of the radiator.
"""
get_name(radiator::Radiator) = String(cpp_radiator_name(radiator._ptr))

"""
    num_height_sections(radiator::Radiator) -> Int

Get the number of sections in the radiator's height grid.
"""
num_height_sections(radiator::Radiator) =
    Int(cpp_radiator_num_height_sections(radiator._ptr))

"""
    num_wavelength_sections(radiator::Radiator) -> Int

Get the number of sections in the radiator's wavelength grid.
"""
num_wavelength_sections(radiator::Radiator) =
    Int(cpp_radiator_num_wavelength_sections(radiator._ptr))

function _radiator_view(address::UInt64, radiator::Radiator)
    dims = (num_height_sections(radiator), num_wavelength_sections(radiator))
    data = unsafe_wrap(Array, Ptr{Float64}(address), dims; own = false)
    return RadiatorView(data, radiator)
end

"""
    optical_depths(radiator::Radiator) -> RadiatorView

Get a zero-copy view of the radiator's optical depths, of shape
`(num_height_sections, num_wavelength_sections)`.

A write to the view changes the radiator.
"""
optical_depths(radiator::Radiator) =
    _radiator_view(cpp_radiator_optical_depths_pointer(radiator._ptr), radiator)

"""
    single_scattering_albedos(radiator::Radiator) -> RadiatorView

Get a zero-copy view of the radiator's single scattering albedos, of shape
`(num_height_sections, num_wavelength_sections)`.

A write to the view changes the radiator.
"""
single_scattering_albedos(radiator::Radiator) =
    _radiator_view(cpp_radiator_single_scattering_albedos_pointer(radiator._ptr), radiator)

"""
    asymmetry_factors(radiator::Radiator) -> RadiatorView

Get a zero-copy view of the radiator's asymmetry factors, of shape
`(num_height_sections, num_wavelength_sections)`.

A write to the view changes the radiator.
"""
asymmetry_factors(radiator::Radiator) =
    _radiator_view(cpp_radiator_asymmetry_factors_pointer(radiator._ptr), radiator)

function _check_radiator_array_size(
    radiator::Radiator,
    values::AbstractMatrix{<:Real},
    label::AbstractString,
)
    expected = (num_height_sections(radiator), num_wavelength_sections(radiator))
    size(values) == expected || error(
        "$label must have size $expected (num_height_sections, num_wavelength_sections).",
    )
end

"""
    set_optical_depths!(radiator::Radiator, values::AbstractMatrix{<:Real}) -> Radiator

Copy `values` into the radiator's optical depths.
"""
function set_optical_depths!(radiator::Radiator, values::AbstractMatrix{<:Real})
    _check_radiator_array_size(radiator, values, "optical_depths")
    cpp_radiator_set_optical_depths!(radiator._ptr, vec(convert(Matrix{Float64}, values)))
    return radiator
end

"""
    set_single_scattering_albedos!(radiator::Radiator, values::AbstractMatrix{<:Real}) -> Radiator

Copy `values` into the radiator's single scattering albedos.
"""
function set_single_scattering_albedos!(radiator::Radiator, values::AbstractMatrix{<:Real})
    _check_radiator_array_size(radiator, values, "single_scattering_albedos")
    cpp_radiator_set_single_scattering_albedos!(
        radiator._ptr,
        vec(convert(Matrix{Float64}, values)),
    )
    return radiator
end

"""
    set_asymmetry_factors!(radiator::Radiator, values::AbstractMatrix{<:Real}) -> Radiator

Copy `values` into the radiator's asymmetry factors.
"""
function set_asymmetry_factors!(radiator::Radiator, values::AbstractMatrix{<:Real})
    _check_radiator_array_size(radiator, values, "asymmetry_factors")
    cpp_radiator_set_asymmetry_factors!(
        radiator._ptr,
        vec(convert(Matrix{Float64}, values)),
    )
    return radiator
end

function Base.show(io::IO, radiator::Radiator)
    print(
        io,
        "Radiator(name=\"$(get_name(radiator))\", ",
        "num_height_sections=$(num_height_sections(radiator)), ",
        "num_wavelength_sections=$(num_wavelength_sections(radiator)))",
    )
end

export Radiator, RadiatorView
export get_name, num_height_sections, num_wavelength_sections
export optical_depths, single_scattering_albedos, asymmetry_factors
export set_optical_depths!, set_single_scattering_albedos!, set_asymmetry_factors!
